File: BaekJoon/Java/23843_콘센트_PriorityQueue_Greedy.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// 콘센트 23843 PriorityQueue + Greedy
// O( N logN )


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        Integer[] times = new Integer[n];
        for (int i = 0; i < n; ++i) {
            times[i] = Integer.parseInt(st.nextToken());
        }

        // 긴 순서대로 내림차순 정렬
        Arrays.sort(times, Collections.reverseOrder());
        
        // 최소 힙
        PriorityQueue<Integer> pq = new PriorityQueue<>(m);
        for (int i = 0; i < m; ++i) {
            pq.add(0);
        }

        int minTime = 0;
        for (int i = 0; i < n; ++i) {
            int cur = pq.poll() + times[i];
            pq.add(cur);

            minTime = Math.max(minTime, cur);
        }

        System.out.println(minTime);
    }
}

File: BaekJoon/23843_콘센트_PriorityQueue_Greedy.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

// 콘센트 23843 PriorityQueue + Greedy
// O( N logN )


int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n, m;
    cin >> n >> m;

    vector<int> times(n);   // 기기별 충전 시간
    for (int& t : times)
    {
        cin >> t;
    }

    // 긴 순서대로 내림차순 정렬
    sort(times.rbegin(), times.rend());

    // 최소 힙
    priority_queue<int, vector<int>, greater<int>> pq;
    for (int i = 0; i < m; ++i) // pq 초기화
    {
        pq.push(0);
    }

    int minTime{ 0 };   // 모든 기기가 충전되는데 걸리는 최소 시간
    for (int i = 0; i < n; ++i)
    {
        // 충전이 가장 빨리 끝나는 콘센트에
        // 가장 오래 걸리는 기기 충전
        int cur = pq.top() + times[i];
        pq.pop();
        pq.push(cur);
        
        minTime = max(minTime, cur);
    }

    cout << minTime;


    return 0;
}
